### PR TITLE
Properly handle removal of REST stat metrics when an application is unloaded

### LINKED ISSUE #10942: Properly remove REST stat metrics when applications are unloaded [mpMetrics-2.3]
REST stat metrics are not properly removed when an application is unloaded/removed.

### PR DESCRIPTION
**Pull request by Channyboy**
#build  
fixes #10942 

REST stat metrics are not properly removed when an application is unloaded/removed.
Uses the SharedMetricRegistry service component to piggy back a call to MetricRegistryImpl

Eclipse formatted the following file dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java

See the comments for listing of what files were changed.

### CODE REVIEW

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 1]
```diff
@@ -29,187 +29,213 @@
...
+			}
+
+			// reset
+			metricRegistry = null;
+		}
 	}
 
 	protected String[] getRESTMBeanStatsTags() {
-		String[] Tags = new String[3];
+		String[] mbeanProperties = new String[3];
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: This method has been modified.

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 2]
```diff
@@ -29,187 +29,213 @@
...
 	private String getMBeanStatsPortName(String portStr) {
 		portStr = portStr.split("=")[1];
-    	String portName = portStr.replaceAll("\"", "");	
-    	return portName;
+		String portName = portStr.replaceAll("\"", "");
+		return portName;
 	}
 
 	public void unregisterMetrics(SharedMetricRegistries sharedMetricRegistry) {
-		MetricRegistry registry = sharedMetricRegistry.getOrCreate(MetricRegistry.Type.VENDOR.getName());
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: This method has been modified.

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 3]
```diff
@@ -29,187 +29,213 @@
...
+				String appName = objName_rest[0];
+				Tag classTag = new Tag("class", objName_rest[1]);
+				Tag methodTag = new Tag("method", objName_rest[2]);
+
+				metricRegistry.register(Metadata.builder().withName(metricName)
+						.withDisplayName(metricData[MappingTable.METRIC_DISPLAYNAME])
+						.withDescription(metricData[MappingTable.METRIC_DESCRIPTION]).withType(type)
+						.withUnit(metricData[MappingTable.METRIC_UNIT]).build(), mst, classTag, methodTag);
+
+				metricID = new MetricID(metricName, classTag, methodTag);
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: These four lines were modified #2

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 4]
```diff
@@ -29,187 +29,213 @@
...
+	protected Set<MetricID> vendorMetricIDs;
+	protected Set<MetricID> baseMetricIDs;
 
 	public MonitorMetrics(String objectName) {
 		this.mbs = ManagementFactory.getPlatformMBeanServer();
 		this.objectName = objectName;
-		this.metricIDs = new HashSet<MetricID>();
-		
+		this.vendorMetricIDs = new HashSet<MetricID>();
+		this.baseMetricIDs = new HashSet<MetricID>();
```
Comment (Channyboy) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: modifed/new

**Comment by Channyboy** [comment]
Eclipse formatted the following file dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java

See the comments for listing of what files were changed.

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 5]
```diff
@@ -29,187 +29,213 @@
...
 	public void unregisterMetrics(SharedMetricRegistries sharedMetricRegistry) {
-		MetricRegistry registry = sharedMetricRegistry.getOrCreate(MetricRegistry.Type.VENDOR.getName());
-		for (MetricID metricID : metricIDs) {
-			boolean rc = registry.remove(metricID);
+		MetricRegistry vendorRegistry = sharedMetricRegistry.getOrCreate(MetricRegistry.Type.VENDOR.getName());
+		MetricRegistry baseRegistry = sharedMetricRegistry.getOrCreate(MetricRegistry.Type.BASE.getName());
+		
+		for (MetricID metricID : vendorMetricIDs) {
+			boolean rc = vendorRegistry.remove(metricID);
+			Tr.debug(tc, "Unregistered " + metricID.toString() + " " + (rc ? "successfully" : "unsuccessfully"));
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: Put debug statement inside `if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()){ }`

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 6]
```diff
@@ -29,187 +29,213 @@
...
+		}
+		
+		for (MetricID metricID : baseMetricIDs) {
+			boolean rc = baseRegistry.remove(metricID);
 			Tr.debug(tc, "Unregistered " + metricID.toString() + " " + (rc ? "successfully" : "unsuccessfully"));
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: Put debug statement inside `if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()){ }`

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 7]
```diff
@@ -29,187 +29,213 @@
...
+						? new MonitorCounter(mbs, objectName, metricData[MappingTable.MBEAN_ATTRIBUTE])
+						: new MonitorCounter(mbs, objectName, metricData[MappingTable.MBEAN_ATTRIBUTE],
+								metricData[MappingTable.MBEAN_SUBATTRIBUTE]);
+
+				metricRegistry.register(Metadata.builder().withName(metricName)
+						.withDisplayName(metricData[MappingTable.METRIC_DISPLAYNAME])
+						.withDescription(metricData[MappingTable.METRIC_DESCRIPTION]).withType(type)
+						.withUnit(metricData[MappingTable.METRIC_UNIT]).build(), mc, metricTag);
+				metricIDSet.add(metricID);
+				Tr.debug(tc, "Registered " + metricID.toString());
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: Put debug statement inside `if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()){ }`

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 8]
```diff
@@ -29,187 +29,213 @@
...
+				MonitorGauge<Number> mg = metricData[MappingTable.MBEAN_SUBATTRIBUTE] == null
+						? new MonitorGauge<Number>(mbs, objectName, metricData[MappingTable.MBEAN_ATTRIBUTE])
+						: new MonitorGauge<Number>(mbs, objectName, metricData[MappingTable.MBEAN_ATTRIBUTE],
+								metricData[MappingTable.MBEAN_SUBATTRIBUTE]);
+				metricRegistry.register(Metadata.builder().withName(metricName)
+						.withDisplayName(metricData[MappingTable.METRIC_DISPLAYNAME])
+						.withDescription(metricData[MappingTable.METRIC_DESCRIPTION]).withType(type)
+						.withUnit(metricData[MappingTable.METRIC_UNIT]).build(), mg, metricTag);
+				metricIDSet.add(metricID);
+				Tr.debug(tc, "Registered " + metricID.toString());
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: Put debug statement inside `if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()){ }`

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 9]
```diff
@@ -29,187 +29,213 @@
...
+				metricRegistry.register(Metadata.builder().withName(metricName)
+						.withDisplayName(metricData[MappingTable.METRIC_DISPLAYNAME])
+						.withDescription(metricData[MappingTable.METRIC_DESCRIPTION]).withType(type)
+						.withUnit(metricData[MappingTable.METRIC_UNIT]).build(), mst, classTag, methodTag);
+
+				metricID = new MetricID(metricName, classTag, methodTag);
+				metricIDSet.add(metricID);
+				sharedMetricRegistry.associateMetricIDToApplication(metricID, appName, metricRegistry);
+
+				Tr.debug(tc, "Registered " + metricID.toString());
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: Put debug statement inside `if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()){ }`

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 10]
```diff
@@ -29,187 +29,213 @@
...
+			} else {
+				Tr.debug(tc, "Failed to register " + metricName + " because of invalid type " + type);
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: Put debug statement inside `if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()){ }`

File: dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java [hunk 11]
```diff
@@ -29,187 +29,213 @@
...
-            	Tags[2] = Tags[2].replaceAll("\\(\\)", "");
-            	//otherwise first bracket becomes underscores
-            	Tags[2] = Tags[2].replaceAll("\\(", "_");
-            	//second bracket is removed
-            	Tags[2] = Tags[2].replaceAll("\\)", "");
-            	break;
-            }
+
+			if (subString.contains("name=")) {
+				mbeanProperties = subString.split("/");
```
Comment (fmhwong) [inline dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: Not sure the format is.  Is it safe to assume you will always have 3 elements?

Comment (Channyboy) [reply dev/com.ibm.ws.microprofile.metrics.2.3.monitor/src/com/ibm/ws/microprofile/metrics/monitor/MonitorMetrics.java]: Yes, it's expected to be:
`name=ApplicationName/fully.qualified.class.name/methodSignature(java.lang.String)`

This is the Mbean property constructed from the JAXRS filter.

Have made a commit with a comment, and renamed the the variable.

**Comment by Channyboy** [comment]
#build 
It didn't run _all_ the expected FATS?

**Comment by Channyboy** [comment]
#run-libby-bot

**Pull request merged**